File: src/shape.py
from dataclasses import dataclass, field
from functools import wraps
import itertools
from typing import Any, Callable, Iterable, Iterator, NewType


@dataclass(frozen=True)
class ShapeParams:
  css_class: str | None = None
  title: str | None = None
  transform: str | None = None

  def as_text(self) -> str:
    # title must be handled separately.
    data = {"class": self.css_class, "transform": self.transform}
    return "".join(
        f" {key}='{value}'" for key, value in data.items() if value is not None)


@dataclass(frozen=True)
class Rect:
  x: float
  y: float
  w: float
  h: float
  params: ShapeParams = ShapeParams()


@dataclass(frozen=True)
class Line:
  x1: float
  y1: float
  x2: float
  y2: float
  params: ShapeParams = ShapeParams()

  @classmethod
  def vertical(cls,
               x: float,
               y1: float,
               y2: float,
               params: ShapeParams | None = None) -> "Line":
    return cls(x, y1, x, y2, params or ShapeParams())

  @classmethod
  def horizontal(cls,
                 x1: float,
                 x2: float,
                 y: float,
                 params: ShapeParams | None = None) -> "Line":
    return cls(x1, y, x2, y, params or ShapeParams())


@dataclass(frozen=True)
class Circle:
  cx: float
  cy: float
  r: float
  params: ShapeParams = ShapeParams()


@dataclass(frozen=True)
class Text:
  text: str
  x: float
  y: float
  params: ShapeParams = ShapeParams()


Shape = Rect | Line | Circle | Text


class ShapeStream:

  def __init__(self, iterable: Iterable[Shape]) -> None:
    self._it = iterable

  def __iter__(self) -> Iterator[Shape]:
    yield from self._it

  def __add__(self, other: Iterable[Shape]) -> "ShapeStream":
    """Allow: stream_a + stream_b"""
    return ShapeStream(itertools.chain(self._it, other))


def shape_generator(
    func: Callable[..., Iterable[Shape]]) -> Callable[..., ShapeStream]:

  @wraps(func)
  def wrapper(*args: Any, **kwargs: Any) -> ShapeStream:
    return ShapeStream(func(*args, **kwargs))

  return wrapper

File: src/ts/shape.ts
import {PointTransformer} from './point_transformer.js';

export interface ShapeParams {
  cssClass?: string;
  title?: string;
  transform?: string;
  fill?: string;
  stroke?: string;
  strokeWidth?: string;
}

export function shapeParamsAsText(params: ShapeParams): string {
  // Explicitly typing the record ensures strict static checking
  const data: Record<string, string | undefined> = {
    "class": params.cssClass,
    "transform": params.transform,
    "fill": params.fill,
    "stroke": params.stroke,
    "stroke-width": params.strokeWidth,
  };

  return Object.entries(data)
    .filter(([, value]) => value !== undefined)
    .map(([key, value]) => ` ${key}='${value}'`)
    .join("");
}

export type PathPoint = {
  pathType: string; x: number; y: number
};

export type Shape =|{
  type: 'rect';
  x: number;
  y: number;
  w: number;
  h: number;
  params: ShapeParams
}
|{
  type: 'line';
  x1: number;
  y1: number;
  x2: number;
  y2: number;
  params: ShapeParams
}
|{
  type: 'circle';
  cx: number;
  cy: number;
  r: number;
  params: ShapeParams
}
|{
  type: 'text';
  text: string;
  x: number;
  y: number;
  params: ShapeParams
}
|{
  type: 'path';
  points: PathPoint[];
  params: ShapeParams
};

export function transformShape(
    shape: Shape, transformer: PointTransformer): Shape {
  switch (shape.type) {
    case 'line': {
      const [tx1, ty1] = transformer(shape.x1, shape.y1);
      const [tx2, ty2] = transformer(shape.x2, shape.y2);
      return {...shape, x1: tx1, y1: ty1, x2: tx2, y2: ty2};
    }
    case 'rect': {
      const [tx1, ty1] = transformer(shape.x, shape.y);
      const [tx2, ty2] = transformer(shape.x + shape.w, shape.y + shape.h);
      return {
        ...shape,
        x: Math.min(tx1, tx2),
        y: Math.min(ty1, ty2),
        w: Math.abs(tx1 - tx2),
        h: Math.abs(ty1 - ty2)
      };
    }
    case 'circle': {
      const [tx, ty] = transformer(shape.cx, shape.cy);
      const [edgeX] = transformer(shape.cx + shape.r, shape.cy);
      return {...shape, cx: tx, cy: ty, r: Math.abs(edgeX - tx)};
    }
    case 'text': {
      const [tx, ty] = transformer(shape.x, shape.y);
      return {...shape, x: tx, y: ty};
    }
    case 'path': {
      const transformedPoints = shape.points.map((p) => {
        const [px, py] = transformer(p.x, p.y);
        return {pathType: p.pathType, x: px, y: py};
      });
      return {...shape, points: transformedPoints};
    }
  }
}